### PR TITLE
Use Portal for Tooltip

### PR DESCRIPTION
**Pull request by jfurrow**
This uses the portal and will work everywhere (even inside a container with `overflow: scroll`). That introduced a fair amount of additional logic to determine the position of the tooltip trigger and the ideal location for the tooltip content.

### CODE REVIEW

**Comment by mesosphere-web** [comment]
Reviewer: @rcorral

File: src/Tooltip/Tooltip.js [hunk 1]
```diff
@@ -4,91 +4,183 @@ import ReactDOM from 'react-dom';
...
+    }
+
+    return {
+      left: triggerRect.left - tooltipWidth + ARROW_SIZE,
+      top: triggerRect.top + triggerRect.height / 2
+    };
+  }
+
+  isVertical(position) {
+    if (position === 'left' || position === 'right') {
```
Comment (kennyt) [inline src/Tooltip/Tooltip.js]: we can just do `return position !== 'left' && position !== 'right'`

File: src/Tooltip/Tooltip.js [hunk 2]
```diff
@@ -4,91 +4,183 @@ import ReactDOM from 'react-dom';
...
 
     return position;
   }
 
-  getIdealLocation() {
-    let anchor = this.state.anchor || this.props.anchor;
-    let position = this.state.position || this.props.position;
+  getIdealLocation(anchor, position) {
+    let isVertical = this.isVertical(position);
+    let triggerNode = ReactDOM.findDOMNode(this.refs.triggerNode);
```
Comment (kennyt) [inline src/Tooltip/Tooltip.js]: i believe `this.refs.triggerNode` will already give you the dom node in this version of react

**Comment by jfurrow** [comment]
@kennyt I implemented the things we discussed on Friday, but made some modifications:
* The scroll listener will only be added when the tooltip is open, and is removed when the tooltip closes, which should prevent any performance issues.
* The user can pass in a `scrollContainer` prop, but it's expected to be a DOM node rather than a selector (like the table accepts). I chose a DOM node because many tooltips will have the same container, so I figured we could save some time by getting the DOM node in the parent component and passing it to each tooltip.
* I created a new DOMUtil called `getNodeClearance` to get the number of pixels around an element.

Let me know what you think!

File: src/Tooltip/Tooltip.js [hunk 3]
```diff
@@ -4,91 +4,173 @@ import ReactDOM from 'react-dom';
...
+    this.addScrollListener();
+  }
+
+  handleMouseLeave() {
+    this.dismissTooltip();
+  }
+
+  handleTooltipMouseEnter() {
+    if (this.props.interactive) {
+      this.setState({isOpen: true});
```
Comment (kennyt) [inline src/Tooltip/Tooltip.js]: do we need to readd the listener here? or does `handleMouseLeave` never get called at this point?

Comment (jfurrow) [inline src/Tooltip/Tooltip.js]: Yeah it should be here also. Done!

**Comment by kennyt** [comment]
you currently have 4 tooltip examples of non interactive. can we do 2 interactive and 2 non interactive?

**Comment by jfurrow** [comment]
@kennyt Added interactive examples and added the ability to pass a string as a selector for the scroll container. I believe this covers everything!

File: src/Tooltip/Tooltip.js [hunk 4]
```diff
@@ -4,91 +4,183 @@ import ReactDOM from 'react-dom';
...
+    return this.transformAnchor(anchor, clearance.top, clearance.bottom,
+      tooltipHeight, clearance.boundingRect.height);
   }
 
-  getPosition(position, renderedPosition, viewportWidth, viewportHeight) {
+  getCoordinates(position, clearance, tooltipWidth, tooltipHeight) {
+    // Calculate the coordinates of the tooltip content.
+    if (position === 'top') {
+      return {
+        left: clearance.boundingRect.left + clearance.boundingRect.width / 2,
```
Comment (kennyt) [inline src/Tooltip/Tooltip.js]: `left` is already on `clearance`, right? so you can just do `clearance.left`

**Comment by kennyt** [comment]
after my one last comment this looks good :ship:

File: src/Tooltip/Tooltip.js (hunk 4 above)
Comment (jfurrow) [inline src/Tooltip/Tooltip.js]: These are actually different values... kind of. The `top`, `right`, `bottom`, and `left` properties directly on the `clearance` object reflect the number of pixels between the edge of the viewport and the edge of the element. The `boundingRect` object contains the position of the element on the screen. So while sometimes these values are the same (`top` and `left` will be the same for both values), their intentions are different.

**Pull request merged**